### PR TITLE
Update ep_calcs.R

### PR DESCRIPTION
**Pull request by spfleming**
Converted lead() code to case-by-case to ensure the next-play down, distance, and yard line were correct.
Converted kickoff down to consistent (down = 5) for easier identification.
Defined 5 categories of plays(Normal, Turnover, Defense Score, Offense Score, Kickoff).

### CODE REVIEW

File: R/ep_calcs.R [hunk 1]
```diff
@@ -370,32 +370,67 @@ prep_df_epa2 <- function(dat){
...
+        play_type %in% normalplay ~ adj_yd_line - yards_gained,
+        play_type %in% score ~ 0,
+        play_type %in% defense_score ~ 0,
+        play_type %in% kickoff ~ start_yards_to_goal,
+        play_type %in% turnover ~ 100 - adj_yd_line + yards_gained),
+      
       new_TimeSecsRem = lead(TimeSecsRem),
       new_log_ydstogo = log(new_yardline),
-      new_Goal_To_Go = lead(Goal_To_Go),
+      new_Goal_To_Go = ifelse(new_yardline <= distance, 1, 0),
```
Comment (meysubb) [inline R/ep_calcs.R]: shouldn't this be realted to new_distance?

File: R/ep_calcs.R [hunk 2]
```diff
@@ -370,32 +370,67 @@ prep_df_epa2 <- function(dat){
...
       # new under two minute warnings
       new_Under_two = new_TimeSecsRem <= 120,
-      end_half_game=0) %>% ungroup() %>%
+      end_half_game=0)) %>%
```
Comment (meysubb) [inline R/ep_calcs.R]: I think you have one too many paranthesis here. 

Should be end_half_game=0)

File: R/ep_calcs.R [hunk 3]
```diff
@@ -370,32 +370,67 @@ prep_df_epa2 <- function(dat){
...
+        play_type %in% score ~ 1,
+        play_type %in% kickoff ~ 1),
+      
+      new_distance = case_when(
+        play_type %in% normalplay & yards_gained >= distance & (adj_yd_line - yards_gained > 10) ~ 10,
+        play_type %in% normalplay & yards_gained >= distance & (adj_yd_line - yards_gained < 10) ~ adj_yd_line,
+        play_type %in% normalplay & yards_gained < distance & down < 3 ~ distance - yards_gained,
+        play_type %in% normalplay & yards_gained < distance & down == 4 & (100 - (adj_yd_line - yards_gained) > 10) ~ 10,
+        play_type %in% normalplay & yards_gained < distance & down == 4 & (100 - (adj_yd_line - yards_gained) < 10) ~ 100-adj_yd_line,
+        play_type %in% turnover & (100 - (adj_yd_line + yards_gained) > 10) ~ 10,
```
Comment (meysubb) [inline R/ep_calcs.R]: can you change all `adj_yd_line` fields to `yards_to_goal`

File: R/ep_calcs.R [hunk 4]
```diff
@@ -370,32 +370,67 @@ prep_df_epa2 <- function(dat){
...
+      down = ifelse(play_type %in% "Kickoff", 5, down), 
+      
+      new_down = case_when(
+        play_type %in% normalplay & yards_gained >= distance ~ 1,
+        play_type %in% normalplay & yards_gained < distance & down < 3 ~ down + 1,
+        play_type %in% normalplay & yards_gained < distance & down == 4 ~ 1,
+        play_type %in% turnover ~ 1,
+        play_type %in% defense_score ~ 1,
+        play_type %in% score ~ 1,
+        play_type %in% kickoff ~ 1),
```
Comment (meysubb) [inline R/ep_calcs.R]: this fails to create a fourth down.

Comment (meysubb) [reply R/ep_calcs.R]: play_type %in% normalplay & yards_gained < distance & down <= 3 ~ down + 1

Comment (meysubb) [reply R/ep_calcs.R]: that should fix it right there ^

File: R/ep_calcs.R (hunk 1 above)
Comment (spfleming) [reply R/ep_calcs.R]: Yep! oversight. Fixed this.

File: R/ep_calcs.R [hunk 5]
```diff
@@ -370,32 +370,67 @@ prep_df_epa2 <- function(dat){
...
       new_log_ydstogo = log(new_yardline),
-      new_Goal_To_Go = lead(Goal_To_Go),
+      new_Goal_To_Go = ifelse(new_yardline <= new_distance, 1, 0),
       # new under two minute warnings
       new_Under_two = new_TimeSecsRem <= 120,
-      end_half_game=0) %>% ungroup() %>%
-    mutate_at(vars(new_TimeSecsRem), ~ replace_na(., 0)) %>% ungroup()
-
+      end_half_game=0)) %>% 
+    mutate_at(vars(new_TimeSecsRem), ~ replace_na(., 0) %>% ungroup()
```
Comment (meysubb) [inline R/ep_calcs.R:432]: mutate_at(vars(new_TimeSecsRem), ~ replace_na(., 0)) 

Missing an extra paranthesis here.

File: R/ep_calcs.R (hunk 2 above)
Comment (meysubb) [reply R/ep_calcs.R:431]: end_half_game=0)

should be this ^

**Comment by meysubb** [comment]
Can you change all `adj_yd_line` to `yards_to_goal`?

**Comment by meysubb** [comment]
`log_ydstogo = log(distance)`

Need the right variable is logged

File: R/ep_calcs.R [hunk 6]
```diff
@@ -370,32 +370,67 @@ prep_df_epa2 <- function(dat){
...
+      
+      new_down = case_when(
+        play_type %in% normalplay & yards_gained >= distance ~ 1,
+        play_type %in% normalplay & yards_gained < distance & down <= 3 ~ down + 1,
+        play_type %in% normalplay & yards_gained < distance & down == 4 ~ 1,
+        play_type %in% turnover ~ 1,
+        play_type %in% defense_score ~ 1,
+        play_type %in% score ~ 1,
+        play_type %in% kickoff ~ 1),
+
```
Comment (meysubb) [inline R/ep_calcs.R:406]: yards_gained = as.numeric(yards_gained),

in a few test examples, yards_gained is sometimes of an integer type and the distance is numeric and this causes problems in case_when. Best to normalize it before the case_when

Comment (meysubb) [reply R/ep_calcs.R:406]: `start_yards_to_goal = as.numeric(start_yards_to_goal),`

Need to do it here as well.

**Comment by meysubb** [comment]
Some additional fixes include 

changing the turnover vector to be named as `turnover_vec` and same thing with `defensive_score` to `defensive_score_vec` 

Those are column names in the dataframe and its causing some problems.

**Comment by meysubb** [comment]
closing this, since I brought your work in and implemented it

**Pull request closed without merging**